check-services.php

```diff
@@ -61,10 +66,68 @@
     }
 }
 
-$sql = 'SELECT * FROM `devices` AS D, `services` AS S WHERE S.device_id = D.device_id ' . $where . ' ORDER by D.device_id DESC';
+if (!isset($query)) {
+    $query = "SELECT * FROM `devices` WHERE `disabled` = 0 $where ORDER BY `device_id` ASC";
+}
+
+foreach (dbFetch($query) as $device) {
+    $sql = 'SELECT * FROM `devices` AS D
+ INNER JOIN `services` AS S ON S.device_id = D.device_id AND D.device_id = '.$device['device_id'].'
+ LEFT JOIN `devices_attribs` as A ON D.device_id = A.device_id AND A.attrib_type = "override_icmp_disable"
+ ORDER by D.device_id DESC;';
+
+    foreach (dbFetchRows($sql) as $service) {
+    // Run the polling function if the associated device is up, "Disable ICMP Test" option is not enabled,
+    // or service hostname/ip is different from associated device
+        if ($service['status'] === "1" || ($service['status'] === '0' && $service['status_reason'] === 'snmp') ||
+        $service['attrib_value'] === 'true' || ($service['service_ip'] !== $service['hostname'] &&
+        $service['service_ip'] !== $service['ip'] )) {
+            // Mark service check as enabled if it was disabled previosly because device was down
+            if ($service['service_disabled'] === "1") {
+                dbUpdate(
+                    array('service_disabled' => '0'),
+                    'services',
+                    '`service_id` = ?',
+                    array($service['service_id'])
+                );
+            }
+            poll_service($service);
+            $polled_services++;
+        } else {
+            d_echo("\nNagios Service - ".$service['service_id']."\nSkipping service check because device "
+                   .$service['hostname']." is down due to icmp.\n");
+            // Mark service check as disabled while device is down and log to eventlog that service check is skipped,
+            // but only if it's not already marked as disabled
+            if ($service['service_disabled'] === "0") {
+                dbUpdate(
+                    array('service_disabled' => '1'),
+                    'services',
+                    '`service_id` = ?',
+                    array($service['service_id'])
+                );
+                log_event(
+                    "Nagios Service - {$service['service_desc']} ({$service['service_id']}) - 
+                    Skipping service check because device {$service['hostname']} is down due to icmp",
+                    $device,
+                    'service',
+                    4,
+                    $service['service_id']
+                );
+            }
+        }
+    } //end service foreach
+
+    echo "#### Start Alerts ####\n";
+    RunRules($device['device_id']);
```

Comment:
I personally don't think running alerting in here is a good thing.

Reply:
I actually added this as a test, because we had alerting problem I mentioned in comment before.

> I made more changes to the check-services.php logic and I think alerting runs more stable with them. Our instance is running for several days with the changes, and no errors like the ones mentioned below happened after implementing them.
> Could not issue critical alert for rule 'Memory CRITICAL' to transport 'mail' Error: You must provide at least one recipient email address. (Memory CRITICAL is a rule fired based on service check result)
> Before, we had the above error firing several times a day.

I was reviewing alerting and saw RunRules is called in poller.php, so I added it in check-services.php too for the test. Also, kinda made sense to me as poller.php and check-services.php were not working synchronized while polling.

I just reviewed the eventlog of our instance. Before I added RunRules to the code, the error above occurred tens of times on every day. After the change, error still occurred on some days in last month, but 1-4 times tops. It helped, but didn't solve the problem :).

Can't quite figure out what happens that triggers the error. I suspect that RunRules triggers at the same time as alerts.php, but clears/deletes the alert after alerts.php started sending mail transport and because of that it doesn't have any contact to send too. By adding RunRules to check-service.php I guess that situation happened less often.

Off course, it's up to the librenms team to decide. I'm primarily a network/system engineer, so not much programming experience

Reply:
Well if it's not even a fix then I think it needs to be removed so that the loop around two SQL queries can be dropped.